src/api/logs.rs

```diff
@@ -0,0 +1,329 @@
+                        .map(|_| ())
+                        .map_err(|_| ())
+                });
+            }
+        }
+        Ok(current_log)
+    }
+}
+
+impl ViewSet for ApiLogs {
+    /// Lists all logs
+    fn list(&self, _req: Request<Body>) -> ResponseFuture {
+        let cfg_read = self.config.read().unwrap();
+        let mut logs: Vec<Log> = Vec::new();
+        for (_, ds) in &cfg_read.log {
+            logs.push(ds.clone());
+        }
+        let items = ListResponse {
+            total: cfg_read.log.len(),
+            next: None,
+            previous: None,
+            results: logs,
+        };
+        Box::new(self.build_response(items))
+    }
+
+    fn create(&self, req: Request<Body>) -> ResponseFuture {
+        let cfg = Arc::clone(&self.config);
+        Box::new(
+            req.into_body()
+                .concat2()
+                .from_err()
+                .and_then(move |entire_body| {
+                    match ApiLogs::parse_create_body(entire_body.to_vec(), cfg.clone()) {
+                        Ok(mut log) => {
+                            let ds_serialized = serde_json::to_string(&log).unwrap();
+                            let log_name = log.clone().name.unwrap();
+
+                            let res = put_object_metabucket(
+                                cfg,
+                                format!("minsql/meta/logs/{}", log_name),
+                                ds_serialized,
+                            )
+                            .then(move |v| match v {
+                                Ok(_) => {
+                                    log.safe();
+                                    Either::A(future::ok(
+                                        Response::builder()
+                                            .header(header::CONTENT_TYPE, "application/json")
+                                            .body(Body::from(serde_json::to_string(&log).unwrap()))
+                                            .unwrap(),
+                                    ))
+                                }
+                                Err(e) => {
+                                    Either::B(future::ok(return_500(&format!("I/O Err: {}", e))))
+                                }
+                            });
+                            Either::A(res)
+                        }
+                        Err(err_resp) => Either::B(future::ok(err_resp)),
+                    }
+                }),
+        )
+    }
+
+    fn retrieve(&self, _req: Request<Body>, pk: &str) -> ResponseFuture {
+        let cfg_read = self.config.read().unwrap();
+        let mut log = match cfg_read.log.get(pk) {
+            Some(ds) => ds.clone(),
+            None => {
+                return Box::new(future::ok(return_404()));
+            }
+        };
+        log.safe();
+        self.build_response(log)
+    }
+
+    fn update(&self, req: Request<Body>, pk: &str) -> ResponseFuture {
```

Comment:
Wonder if we actually need this API. Why not just do create log as a PUT, make it idempotent and just replace the value if it already exists?

Reply:
Simple completeness of REST, POST creates, PUT updates